### PR TITLE
Upgrade environment variable logic using new Pipeline API for logtail environment configuration management

### PR DESCRIPTION
**Pull request by quzard**
This commit introduces an enhanced file discovery mechanism by transitioning from the old configuration API to the new Pipeline configuration API for ilogtail. With this upgrade, users will experience a change in the way recursive directory scanning is handled.

Previously, the recursive discovery of files within a directory was implied. However, to align with the new API's capabilities and to provide users with explicit control, the recursive behavior must now be explicitly specified in the path configuration.

To enable recursive file discovery within a directory, users should add `/**/` at the point in the path where recursion is desired. This syntax will trigger a recursive scan with a maximum depth of 10 levels. If a non-recursive file discovery is required, users should use a single `/` in the path configuration.

Example usage:
- To recursively discover files: `/var/log/myapp/**/app.log`
- For non-recursive discovery: `/var/log/myapp/app.log`

Please note that if you rely on the recursive file discovery feature, you'll need to update your configuration to continue using this functionality with the new Pipeline API.

### CODE REVIEW

File: pkg/helper/docker_center.go [hunk 1]
```diff
@@ -194,15 +194,16 @@ func (info *K8SInfo) Merge(o *K8SInfo) {
 
 // IsMatch ...
 func (info *K8SInfo) IsMatch(filter *K8SFilter) bool {
```
Comment (messixukejia) [inline pkg/helper/docker_center.go:196]: 这个应该是比较高频的操作，直接加互斥锁，是不是会影响效率？
改成读写锁？

File: helper/envconfig/aliyun_log_config_define.go [hunk 2]
```diff
@@ -22,7 +22,9 @@ import (
 	"strings"
 	"time"
 
-	aliyunlog "github.com/aliyun/aliyun-log-go-sdk"
+	"github.com/alibabacloud-go/tea/tea"
```
Comment (messixukejia) [inline helper/envconfig/aliyun_log_config_define.go:25]: 第一次开发容器相关的代码，需要给下设计，找@太业 重点reveiw下测试设计

File: helper/envconfig/aliyun_log_config_define.go [hunk 3]
```diff
@@ -204,54 +191,55 @@ func initDockerStdoutConfig(dockerInfo *helper.DockerInfoDetail, config *AliyunL
...
+		input = map[string]interface{}{
+			"Type":                     "input_file",
+			"FilePaths":                []string{logPath + "/**/" + filePattern},
+			"EnableContainerDiscovery": dockerFile,
+			"ContainerFilters": map[string]map[string]interface{}{
+				"IncludeEnv": {
+					*flags.LogConfigPrefix + tea.StringValue(config.LogtailConfig.ConfigName): filePath,
+				},
 			},
+			"MaxDirSearchDepth": 20,
```
Comment (messixukejia) [inline helper/envconfig/aliyun_log_config_define.go:221]: MaxDirSearchDepth 这个为什么要加上？

File: helper/envconfig/aliyunlog_client.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+	lock               sync.Mutex
+	lastFetch          time.Time
+	lastRetryFailCount int
+	lastRetryInterval  time.Duration
+	ctx                context.Context
+}
+
+func (c *TokenAutoUpdateClient) flushSTSToken() {
+	for {
```
Comment (messixukejia) [inline helper/envconfig/aliyunlog_client.go:37]: 这个代码很关键，要确保能够生效，测试需要测下超时后是否可以真正生效。

File: helper/envconfig/product_logtail_config.go [hunk 5]
```diff
@@ -24,60 +24,51 @@ import (
...
 	}
 	nginxIngressProcessorDetail := make(map[string]interface{})
 	err := json.Unmarshal(([]byte)(nginxIngressProcessor), &nginxIngressProcessorDetail)
 	if err != nil {
 		logger.Error(context.Background(), "LOGTAIL_CONFIG_ALARM", "invalid nginx ingress config", "processor type error")
 		return
 	}
-	pluginConfigDetail["processors"] = []interface{}{nginxIngressProcessorDetail}
-	logger.Debug(context.Background(), "init nginx ingress config", pluginConfig)
+	config.LogtailConfig.Processors = append(config.LogtailConfig.Processors, nginxIngressProcessorDetail)
```
Comment (messixukejia) [inline helper/envconfig/product_logtail_config.go:70]: nginxingress场景是否测试过

File: helper/envconfig/aliyunlog_client.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+	accessKeyID, accessKeySecret, securityToken, expireTime, err := c.tokenUpdateFunc()
+	if err == nil {
+		c.lock.Lock()
+		c.lastRetryFailCount = 0
+		c.lastRetryInterval = time.Duration(0)
+		c.nextExpire = expireTime
+		c.lock.Unlock()
+		logClient, err := CreateNormalInterface(*c.Client.Endpoint, accessKeyID, accessKeySecret, securityToken, *c.Client.UserAgent)
+		if err != nil {
+			return err
```
Comment (messixukejia) [inline helper/envconfig/aliyunlog_client.go:110]: 假设出了问题，是否有足够信息排查问题？

File: helper/envconfig/aliyunlog_operation_wrapper.go [hunk 7]
```diff
@@ -699,80 +728,80 @@ func (o *operationWrapper) updateConfigInner(config *AliyunLogConfigSpec) error
...
 		machineGroup = config.MachineGroups[0]
 	} else {
+		// 如果配置中没有机器组,使用默认的机器组
 		machineGroup = *flags.DefaultLogMachineGroup
 	}
-	_ = o.makesureMachineGroupExist(project, machineGroup)
-	if ok {
-		var machineGroups []string
-		for i := 0; i < *flags.LogOperationMaxRetryTimes; i++ {
-			machineGroups, err = o.logClient.GetAppliedMachineGroups(project, config.LogtailConfig.ConfigName)
```
Comment (messixukejia) [inline helper/envconfig/aliyunlog_operation_wrapper.go:749]: GetAppliedMachineGroups 这个没有了？

**Review by messixukejia — CHANGES_REQUESTED** [review]

File: helper/envconfig/aliyunlog_operation_wrapper.go (hunk 7 above)
Comment (quzard) [reply helper/envconfig/aliyunlog_operation_wrapper.go:749]: 这块逻辑没啥用，不管GetAppliedMachineGroups结果是啥，都会进行apply

File: helper/envconfig/aliyun_log_config_define.go [hunk 8]
```diff
@@ -204,54 +191,55 @@ func initDockerStdoutConfig(dockerInfo *helper.DockerInfoDetail, config *AliyunL
...
-		config.LogtailConfig.LogtailConfig = map[string]interface{}{
-			"logType":     "common_reg_log",
-			"logPath":     logPath,
-			"filePattern": filePattern,
-			"dockerFile":  dockerFile,
-			"dockerIncludeEnv": map[string]string{
-				*flags.LogConfigPrefix + config.LogtailConfig.ConfigName: filePath,
+		input = map[string]interface{}{
+			"Type":                     "input_file",
+			"FilePaths":                []string{logPath + "/**/" + filePattern},
```
Comment (henryzhx8) [inline helper/envconfig/aliyun_log_config_define.go:214]: 直接把配置的路径放过来，如果有**，那么设置depth和之前一样，如果没有**，则depth为0。具体兼容性自己再检查一下

File: helper/envconfig/aliyun_log_config_define.go [hunk 9]
```diff
@@ -414,13 +403,12 @@ func makeLogConfigSpec(dockerInfo *helper.DockerInfoDetail, envConfigInfo *helpe
 	// config
 	// makesure exist
 	configType := envConfigInfo.ConfigItemMap[""]
-	config.LogtailConfig.LogtailConfig = make(map[string]interface{})
+	// 疑问：新版配置需要支持detail吗
```
Comment (henryzhx8) [inline helper/envconfig/aliyun_log_config_define.go:406]: 把这块删了吧

File: helper/envconfig/product_logtail_config.go (hunk 5 above)
Comment (quzard) [reply helper/envconfig/product_logtail_config.go:70]: k8s-ingress-nginx这个product是不是废弃了？我用老的logtail
使用这个env也创建失败。
aliyun_logs_k8s-ingress-nginx_product=k8s-ingress-nginx

报错提示
ErrorCode: UnRegisterProduct
Message: The product {k8s-ingress-nginx} is not register

File: pkg/helper/docker_center.go [hunk 10]
```diff
@@ -194,16 +194,15 @@ func (info *K8SInfo) Merge(o *K8SInfo) {
...
-	info.mu.Lock()
-	defer info.mu.Unlock()
-
 	if info.PausedContainer {
```
Comment (messixukejia) [inline pkg/helper/docker_center.go]: 改成这样不有变成锁不严格了吗

File: helper/envconfig/aliyun_ecs_meta_data.go [hunk 11]
```diff
@@ -68,7 +68,7 @@ func getToken() (result []byte, err error) {
 		Timeout: time.Second * 3,
 	}
 	var respList *http.Response
-	logger.Debug(context.Background(), "get role list request", aliyunECSRamURL)
+	logger.Info(context.Background(), "get role list request", aliyunECSRamURL)
```
Comment (messixukejia) [inline helper/envconfig/aliyun_ecs_meta_data.go:71]: 这么个打印法，又走向另外一个极端了。自己的判断是什么？

Comment (quzard) [reply helper/envconfig/aliyun_ecs_meta_data.go:71]: 方便判断用了哪个role的token

Comment (messixukejia) [reply helper/envconfig/aliyun_ecs_meta_data.go]: 不能为了方便过度打日志。

File: helper/envconfig/aliyun_log_config_define.go [hunk 12]
```diff
@@ -204,54 +191,62 @@ func initDockerStdoutConfig(dockerInfo *helper.DockerInfoDetail, config *AliyunL
...
 
 // nolint: unparam
 func initFileConfig(k8sInfo *helper.K8SInfo, config *AliyunLogConfigSpec, filePath string, jsonFlag, dockerFile bool) {
-	config.LogtailConfig.InputType = "file"
+	config.LogtailConfig.Inputs = make([]map[string]interface{}, 0, 1)
+
+	// 判断有没有 /**/
+	maxDirSearchDepth := 0
+	if strings.Contains(filePath, "/**/") {
+		maxDirSearchDepth = 100
```
Comment (yyuuttaaoo) [inline helper/envconfig/aliyun_log_config_define.go:209]: 10

File: helper/envconfig/docker_env_config_manager.go [hunk 13]
```diff
@@ -69,11 +69,11 @@ func (decm *Manager) removeUselessCache() {
 func (decm *Manager) run() {
 	decm.shutdown = make(chan struct{})
 	var err error
-	var clientInterface aliyunlog.ClientInterface
+	var logClient **aliyunlog.Client
```
Comment (yyuuttaaoo) [inline helper/envconfig/docker_env_config_manager.go:72]: 2层*能不能避免一下？

File: helper/envconfig/aliyun_log_config_define.go (hunk 3 above)
Comment (quzard) [reply helper/envconfig/aliyun_log_config_define.go]: 因为默认0，现在需要规定，path里有/**/才会设置maxDirSearchDepth = 10

File: helper/envconfig/aliyunlog_operation_wrapper.go [hunk 14]
```diff
@@ -713,14 +708,18 @@ func (o *operationWrapper) updateConfigInner(config *AliyunLogConfigSpec) error
...
 						needUpdate = true
 						break
 					}
 				}
 			}
-			logger.Info(context.Background(), "config updated, needUpdate", needUpdate, "server config", *serverConfig, "local config", *config)
+			if needUpdate {
+				logger.Info(context.Background(), "config updated, needUpdate", needUpdate, "server config", *serverConfig, "local config", *config)
+			} else {
+				logger.Info(context.Background(), "config not changed", "skip update")
```
Comment (yyuuttaaoo) [inline helper/envconfig/aliyunlog_operation_wrapper.go:721]: Debug

File: helper/envconfig/aliyunlog_operation_wrapper.go [hunk 15]
```diff
@@ -752,6 +751,7 @@ func (o *operationWrapper) updateConfigInner(config *AliyunLogConfigSpec) error
 		} else {
 			o.eventRecorder.SendNormalEventWithAnnotation(o.eventRecorder.GetObject(), annotations, k8s_event.UpdateConfig, "update config success")
 		}
+		logger.Info(context.Background(), "config created, config", *config)
```
Comment (yyuuttaaoo) [inline helper/envconfig/aliyunlog_operation_wrapper.go:754]: 和719行统一一下

**Review by yyuuttaaoo — APPROVED** [review]

**Pull request merged**